### ISSUE TITLE
[cc65] Address size qualifiers?

### ISSUE THREAD

**Issue by acqn** [cc65/cc65#1162]
There are "address size" qualifiers for code/data pointers in cc65, but for now only `__near__` and `__far__` are supported, one of which will be automatically added to pointers at definitions according to the memory model if neither is explicitly specified. The question is, what is the consequence if a pointer does have no such qualifiers at all? This can easily happen as an array is converted to a pointer, which could also expose some hidden type comparison issues .

Example:
```
void f(int*);   /* the parameter will have '__near__' by default */
void f(int []); /* the parameter will have no qualifiers at all when converted to a 'int *' pointer */
```
The two function prototypes are considered mismatched even if the array parameter is converted to pointer due to the first having an `__near__` qualifier on the parameter while the second having none after conversion.

**Comment by mrdudz** [comment]
everything that isnt explicitly "far" should become "near", IMHO

**Comment by polluks** [comment]
Indeed, "near" should be shorter and shorter should be default anyway.

**Comment by acqn** [comment]
But... I suppose "no address size qualifier" should be equivalent to the "default", and the "default" should be the same as the memory model specifies?

**Comment by mrdudz** [comment]
there is a memory model that has anything different than "near" as default?

**Comment by oliverschmidt** [comment]
> I suppose "no address size qualifier" should be equivalent to the "default", and the "default" should be the same as the memory model specifies?

I'm afraid that there's a misunderstanding. I think that @mrdudz, @polluks and me can't imagine anything else beside that approach. So we don't really understand what question or decision we are discussing here. Or it's just me missing the point...

**Comment by acqn** [comment]
https://github.com/cc65/cc65/blob/0c72647edda327df6970773cefb28d3c7c6a41d9/src/common/mmodel.c#L91-L126

**Comment by acqn** [comment]
> 
> 
> there is a memory model that has anything different than "near" as default?

https://github.com/cc65/cc65/blob/0c72647edda327df6970773cefb28d3c7c6a41d9/src/common/mmodel.c#L108-L118

Yes, there is.

**Comment by mrdudz** [comment]
And it's actually used by any of the targets? IIRC the compiler cant even deal with far pointers? *shrug*

**Comment by greg-king5** [comment]
Uz wanted to add support, to the compiler, for the 65816 CPU.  He added the keywords, but almost nothing else.  `__far__` does nothing.  Therefore, `DoCompare()` should ignore the address-size qualifiers -- they don't make different objects!

**Comment by oliverschmidt** [comment]
@acqn: Could you please comment on the status of this issue from your POV?

**Comment by acqn** [comment]
The example in the OP now works since PR #1218 adds the qualifier in parameter conversions, but the issue hasn't been technically fixed. I can work on it.

**Comment by oliverschmidt** [comment]
Thanks for the feedback - and the offer of course!

**Comment by acqn** [comment]
Is a `__far__` pointer supposed to be a 32-bit pointer for cc65? If so, the `__far__` address size "qualifier" would change the representation of the pointer that is 16-bit, and would not be a comformant to the C standard for a qualifier shall not change the representation of a type.

**Comment by mrdudz** [comment]
i think they were never really specified.... It was ment for 65816 support, so Uz might just as well have thought of 24bits :)

In any case, the representation has to be different. I wonder how eg 16bit DOS compilers handled this? Standards werent much of a big thing back then, right? :)

**Comment by oliverschmidt** [comment]
I'm almost sure that e.g. in Windows 3.x, near pointers were 16 bit and far pointers were 32 bit.

BTW: Beside those there were huge pointers (32 bit) and based pointers (16 bit).

**Comment by groessler** [comment]
On 11/15/22 16:57, Oliver Schmidt wrote:
>
> I'm almost sure that e.g. in Windows 3.x, near pointers were 16 bit 
> and far pointers were 32 bit.
>
> BTW: Beside those there were huge pointers (32 bit) and based pointers 
> (16 bit).
>

And, if you remember Watcom, in its 32bit mode far and huge pointers 
were 48bit. 16bit segment and 32bit offset.

regards,
chris

**Comment by spiro-trikaliotis** [comment]
These near, far, huge pointers (in 16 bit and 32 bit mode) were exactly the pointer type that the underlying hardware, the CPU, had, so it is not a big surprise.

Near pointers were 16 bit or 32 bit (depending on 16 or 32 bit mode), while far pointers added another 16 bit (for the segment). Huge pointers were essentially the same as far pointers, the only difference was that it allowed calculations / adjustments on the segments' part, if needed.

QNX 2.x/3.x handled it a little bit differently, but I don't remember the details at the moment. I cannot find the handbook, either.

BTW, these were non-standard extensions to compilers that predated ANSI, so it does not matter that they violate the standard.

**Comment by pm100** [comment]
my 10c worth. 
- cc65 help says it supports short and long options for memory model. It only actually supports the long options
- there is no indication in the docs what values are allowed and what they mean
- "huge" is a syntactically valid option but is unconditionally rejected
- as far as I can see compiling with near or far produces identical code

**Comment by kugelfuhr** [comment]
Could someone please tell me what this bug report is about? The issue seems to be fixed, the example code works. So what exactly is missing now?

It seems that the `--memory-model` option to cc65 doesn't have an effect. So is this PR about removing it completely? Not sure about that. It seems to be an unfinished feature.

**Comment by mrdudz** [comment]
> It seems to be an unfinished feature.

It is. Its much in a similar state as the floating point stuff - see Gregs comment:

> Uz wanted to add support, to the compiler, for the 65816 CPU. He added the keywords, but almost nothing else.

I am not sure what would be the right thing to do here. On one side i don't like those unfinished things being there and want to remove them - on the other hand the stuff Uz prepared is usually just fine and can be built upon, so leaving it there makes sense. (Clearly artefacts from the times when developing in feature branches wasn't really popular. Or something)

I probably labelled this "bug" (*) because the docs don't match the behaviour.

My personal favourite solution for this would be to do as we do in VICE - we have a global define ENABLE_EXPERIMENTAL_FEATURES which will enable stuff like this (and is disabled in normal builds). I have no idea how much of a practical solution this would be for this project (especially for the compiler, where code for features is often scattered around... i don't want to create an ifdef hell either)

In any case, it would probably be a good idea to just point this out in the documentation for the time being

(*) mind you that happend long after the ticket was created, we didn't have labels before. This is also the reason for why there are so many issues, some labelled imprecise, some very similar to others, etc. When i took over there were several hundred unlabelled issues, sometimes years old, and often i had to guess, quickly (going through them still took more time than i thought). And yes, just like you i find this a bit annoying and sometimes overwhelming - it is already much better than before though, at least you can filter them somewhat now.

**Comment by kugelfuhr** [comment]
Using something like `ENABLE_EXPERIMENTAL_FEATURES` won't work here since it requires "someone" to look through all of the code identifying every piece that is part of the unfinished feature and `#ifdef` it. I don't know of anybody willing to do that. Besides that, code `#ifdef`ed out tends to rot since it isn't compiled so after some time no one knows if it works any more. And, last thing, `#if`s make the code unreadable.

So do we agree that fixing this issue requires to update the documentation so it mirrors the actual behavior?

**Comment by mrdudz** [comment]
> So do we agree that fixing this issue requires to update the documentation so it mirrors the actual behavior?

yes, totally

**Comment by kugelfuhr** [comment]
Just for the docs:
* Memory model `near` means 16 bit pointer for code and data.
* Memory model `far` means 24 bit pointers for code and 16 bit pointers for data.
* Memory model `huge` means 24 bit pointers for code and data.

Taken from `common/mmodel.c`. So `far` and `huge` seem to be targeted to the 65816. There is no memory model that uses 32 bit addresses.

The address sizes enabled with `--memory-model` will only make it into the generated types but don't have any further effect when generating code. So it does indeed seem like an unfinished feature.
